### PR TITLE
Add compaction policy files

### LINKED ISSUE #3362: add LevelCompactionPolicy for new compaction execution framework
## Enhancement
 LevelCompactionPolicy is created based on CompactionContext to decide whether the tablet should be compacted, which level should be compacted, and how many rowsets in that level should be compacted. And LevelCompactionPolicy use the same strategy as before. The details are as follows: when the compaction score is above threshold, it is considered to be compaction candidate. It use the same strategy to pick rowsets from level. But the difference is there are no cumulative compaction or base compaction. Now it replaced by level compaction. And there are only three levels and also two types of compaction(level-0 and level-1 compaction) to keep compatible. Level-0 compaction is equivalent to cumulative compaction as before; Level-1 compaction is equivalent to base compaction. Maybe it can be extended to more levels in the future.

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [ ] bug
- [X] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3362

## Problem Summary(Required) ：

Add BaseAndCumulativeCompactionPolicy for new compaction execution framework. BaseAndCumulativeCompactionPolicy is created based on CompactionContext to decide whether the tablet should be compacted, which type of compaction should be done, and how many rowsets should be compacted. And BaseAndCumulativeCompactionPolicy uses the same strategy as before. The details are as follows: when the compaction score is above threshold, it is considered to be compaction candidate. It use the same strategy to pick rowsets as before.

### CODE REVIEW

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/storage/level_compaction_policy.cpp [hunk 1]
```diff
@@ -221,7 +221,7 @@ double LevelCompactionPolicy::_get_base_compaction_score() {
         level_1_rowsets_size += rowset->data_disk_size();
     }
 
-    double num_score = static_cast<double>(segment_num_score) / config::base_compaction_num_cumulative_deltas;
+    double num_score = static_cast<double>(segment_num_score) / config::min_base_compaction_num_singleton_deltas;
     double size_score = static_cast<double>(level_1_rowsets_size) / config::min_base_compaction_size;
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply be/src/storage/level_compaction_policy.cpp:225]: compaction_policy.h is for interface. I rename level_compaction_policy to original_compaction_policy

File: be/src/storage/compaction_utils.cpp [hunk 2]
```diff
@@ -117,8 +116,8 @@ CompactionAlgorithm CompactionUtils::choose_compaction_algorithm(size_t num_colu
 }
 
 std::unique_ptr<CompactionPolicy> CompactionUtils::create_compaction_policy(CompactionContext* context) {
-    // now only support LevelCompactionPolicy
```
Comment (chaoyli) [inline be/src/storage/compaction_utils.cpp:120]: Original is for what?

**Review by meegoo — DISMISSED** [review]

File: be/src/storage/compaction_utils.cpp (hunk 2 above)
Comment (ABingHuang) [reply be/src/storage/compaction_utils.cpp:120]: the compaction policy is the same as original compaction strategy. So I use original.

File: be/src/storage/level_compaction_policy.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ABingHuang) [reply be/src/storage/level_compaction_policy.cpp:226]: change it to base_cumulative_compaction_policy

File: be/src/storage/compaction_utils.cpp (hunk 2 above)
Comment (ABingHuang) [reply be/src/storage/compaction_utils.cpp:120]: change it to base_cumulative

**Review by meegoo — DISMISSED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — DISMISSED** [review]

**Comment by chaoyli** [comment]
run starrocks_clang-format

File: be/src/storage/CMakeLists.txt [hunk 3]
```diff
@@ -153,4 +153,5 @@ add_library(Storage STATIC
     compaction_manager.cpp
     compaction_scheduler.cpp
     compaction_task_factory.cpp
+    base_cumulative_compaction_policy.cpp
```
Comment (chaoyli) [inline be/src/storage/CMakeLists.txt:156]: 1. base_and_cmulative_compaction_policy.cpp
2. The commit message should be changed also.

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Comment by decster** [comment]
Add BaseAndCumulativeCompactionPolicy for new compaction execution framework. BaseAndCumulativeCompactionPolicy is created based on CompactionContext to decide whether the tablet should be compacted, which type of compaction should be done, and how many rowsets should be compacted. And BaseAndCumulativeCompactionPolicy uses the same strategy as before. The details are as follows: when the compaction score is above a threshold, it is considered to be a compaction candidate. It uses the same strategy to pick rowsets as before.

**Pull request merged**